tools/eval/datasets/minisupervisely.py

```diff
@@ -0,0 +1,206 @@
+import os
+import cv2 as cv
+import numpy as np
+from tqdm import tqdm
+
+
+class MiniSupervisely : 
+    def __init__(self, root) : 
+        self.root = root
+        self.val_path = os.path.join(root, 'val.txt')
+        self.image_set = self.load_data(self.val_path)
+        self.num_classes = 2
+        self.miou = -1
+        self.class_miou = -1
+        self.acc = -1
+        self.class_acc = -1
+
+
+    @property
+    def name(self):
+        return self.__class__.__name__
+    
+
+    def load_data(self, val_path) : 
+        """
+        Load validation image set from val.txt file
+        Args :
+            val_path (str) : path to val.txt file
+        Returns :
+            image_set (list) : list of image path of input and expected image
+        """
+
+        image_set = []
+        with open(val_path, 'r') as f : 
+            for line in f.readlines() : 
+                image_set.append(line.strip().split())
+
+        return image_set
+    
+    
+    def eval(self, model) : 
+        """
+        Evaluate model on validation set
+        Args :
+            model (object) : PP_HumanSeg model object
+        """
+
+        intersect_area_all = []
+        pred_area_all = []
+        label_area_all = []
+
+        pbar = tqdm(self.image_set)
+
+        for input_image, expected_image in pbar : 
+            pbar.set_description(
+                "Evaluating {} with {} val set".format(model.name, self.name))
+            
+            input_image = cv.imread(os.path.join(self.root, input_image))
+            input_image = cv.resize(input_image, (192, 192))
```

Comment:
Missing two transforms:

1. to_rgb: https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/transforms/transforms.py#L61-L62. Note that to_rgb is set to True by default.
2. normailze: 
    1. Config: https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/contrib/HumanSeg/configs/fcn_hrnetw18_small_v1_humanseg_192x192_mini_supervisely.yml#L20-L28 
    2. Normalize transform: https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/transforms/transforms.py#L402-L444

Reply:
I dont see a direct transforms applied in the [infer method](https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/core/infer.py#L192) used by the [evaluation method](https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/core/val.py#L28).  The [inference script](https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/legacy/contrib/HumanSeg/infer.py#L68) of the model in PaddleSeg uses it. So Im not sure if there's a need for it in this eval script

Reply:
> mIoU has decreased to 0.3562 .

My suggestion would be doing alignment with PaddleSeg's results image by image. For example, you could add `break` right after `calculate_area` in your code and PaddleSeg, print `intersect_area, pred_area, label_area` to see whether your results are completely the same with PaddleSeg's. You will also need to ensure you are comparing the same image since PaddleSeg uses a data loader to load images which may have shuffled images.

Reply:
@fengyuentau The reason mIoU decresead to 0.3562 was doing normalization twice. [model.infer](https://github.com/opencv/opencv_zoo/blob/6c68bc48c6f96042b29b3425174e431ccac38376/models/human_segmentation_pphumanseg/pphumanseg.py#L44) of PPHumanSeg already does normalization before passing the input to model so we dont have to do normalization again in the evaluation script. I've reverted that back.

The reason behind why current eval script gives mIoU of 0.8798 and not the expected 0.8602 is - model outputs a 192x192 tensor(logits). Original PaddleSeg script resizes this tensor back to original image size before taking one-hot of that tensor and calculating accuracy, IoU (all calculations done on tensor of shape equal to input image).
This can be seen below : 
[infer method called by PaddleSeg evaluation script](https://github.com/PaddlePaddle/PaddleSeg/blob/13aac35b31b147d2dd9976821ade13dabebff685/paddleseg/core/infer.py#L134)
![image](https://user-images.githubusercontent.com/94868003/219100036-92a290d7-71c9-4c7b-b64a-9f2631524153.png)
Here's the output (the input image is of size 800x1313 in this case) : 
![image](https://user-images.githubusercontent.com/94868003/219100458-7d8112ef-f2cf-4ad1-a5fb-df05dcdeaa93.png)

On the other hande, current eval script resizes the expected_image (label) to 192x192 (same as output of model) and then calculates accuracy, IoU. 
This is because the infer method of PPHumanSeg does not resize the input back to its original shape and directly calculates the one-hot. And thus all the calculation (for mIoU and accuracy) is done on a 192x192 tensor and not the original image size tensor.
![image](https://user-images.githubusercontent.com/94868003/219110459-403907ce-302f-4548-a8b5-42721b94e8a5.png)

![image](https://user-images.githubusercontent.com/94868003/219110336-c33b4a66-3adc-4624-b258-1b2c5a29ea34.png)


Should I proceed with modifying the infer method of PPHumanSeg class to resize the output back to original image shape before calculating the one-hot?

Reply:
> Should I proceed with modifying the infer method of PPHumanSeg class to resize the output back to original image shape before calculating the one-hot?

No, the idea of infer method of PPHumanSeg is assuming the input image is resized to the target size [192, 192] already. Thus, any resize operation should be done here, say right after `output_image = model.infer(input_image)   ` we `reverse_transform(output_image)`.

`to_rgb` needs to be kept or directly add this in `PPHumanSeg._preprocess`.

Reply:
The logits output of model.infer should be resized. model.infer returns the one-hot and not the logits.

so `output_image` is a one-hot and resizing that won't be the same as resizing the `logits` of the model.(correct me if Im wrong).

If we want to do `reverse_transform(outpu_image)` in the eval script, we should change the model.infer method to return the logits and not postprocess it to get the one-hot.

So
- resize the logits in model.infer method before calculating one-hot(postprocess)
- return the logits from model.infer and do reverse_transform in the eval scrirpt.

which one should we proceed with?

Reply:
Okay. I propose to add resize (to target_size [192, 192]) in `_preprocess` and reverse resize (to its original size) in `_postprocess` in `PPHumanSeg` instead.

Reply:
I made the above changes, however the mIoU now is 0.9085 .

I started by comparing the 2 models. The [human_segmentation_pphumanseg_2021oct.onnx](https://github.com/opencv/opencv_zoo/blob/master/models/human_segmentation_pphumanseg/human_segmentation_pphumanseg_2021oct.onnx) downloaded from [this repository](https://github.com/opencv/opencv_zoo/tree/master/models/human_segmentation_pphumanseg) was evaluated on the script I've written so far. 
The [PP-HumanSegV1-Lite](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.7/contrib/PP-HumanSeg#42-general-human-segmentation-models) model downloaded from PaddleSeg repo (branch release/2.7). It has mUoU of 0.8602 . It was evaluated using the [eval script](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.7/contrib/PP-HumanSeg#63-evaluation) provided by PaddleSeg

First I compared the input to both models (make sure preprocessing is done properly) : 
zoo model : 
![image](https://user-images.githubusercontent.com/94868003/219444281-01a9ef1d-24fe-445e-a4b9-a7cd2c2d1367.png)
PaddleSeg model ([infer method](https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/core/infer.py#L192) called by eval script): 
![image](https://user-images.githubusercontent.com/94868003/219444847-d5ffab92-40c3-450f-a13e-871fcacff3c0.png)

Both the tensors were equivalent : 
![image](https://user-images.githubusercontent.com/94868003/219444967-b4a28909-23f4-4ddc-ba45-601d44dababe.png)


Next I compared the immediate outputs of the model : 
zoo model : 
![Screenshot from 2023-02-16 23-15-02](https://user-images.githubusercontent.com/94868003/219446223-4328cf7d-a45d-4ff9-a079-598089a08960.png)
PaddleSeg model : 
![image](https://user-images.githubusercontent.com/94868003/219446323-20c0ad80-1d2b-46cc-b7b1-55effab264cd.png)

Surprisingly these 2 outputs were different!
![Screenshot from 2023-02-16 23-15-23](https://user-images.githubusercontent.com/94868003/219446404-1b4653f2-c2c8-4424-b768-14ce1905201a.png)

(I made sure both of them were using the same input image).

Most likely Im using a different model for comparision? Let me know where can I find the actual model or if Im doing something else wrong

Reply:
> Most likely Im using a different model for comparision?

It could happen. The one here in the zoo is converted from Paddle to ONNX using [this script](https://github.com/opencv/opencv/tree/master/samples/dnn/dnn_model_runner/dnn_conversion/paddlepaddle#2-run-paddleseg-portrait-segmentation-using-opencv), and the script has been there for 2 years. Note that the model was provided in a link which is different from all other links in PaddleSeg:release/2.7. Paddle could have updated this model during this time and somehow still name it "v1".

It can be verified with using your evaluation script to evaluate the one from PaddleSeg:release/2.7. Paddle provides a conversion tool https://github.com/PaddlePaddle/Paddle2ONNX. If you can reach the same mIoU (0.8602) with your evaluation script and the model from PaddleSeg:release/2.7, that's a success.

Reply:
I tried to convert the models from [PaddleSeg](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.7/contrib/PP-HumanSeg#42-general-human-segmentation-models) into onnx format.

the PP-HumanSegV1-Lite (checkpoint) gives the following error
![image](https://user-images.githubusercontent.com/94868003/219581274-061ebb1f-cb9f-42c9-991c-47846c11fa38.png)

I tried other models (v2,mobile) from there, none of the checkpoint models were converted to onnx (same error).

then I tried the Inference Model(softmax) of v1. It was converted to onnx successfully.
However, `cv.dnn.readNet()` wasn't able to read it : 
![image](https://user-images.githubusercontent.com/94868003/219590152-4d9774d4-5291-4d91-923b-d4ebbff1e7b1.png)

The readNet function gave the same error for Mobile inference model(softmax).

At this point, PaddleSeg doesn't provide the model currently in this repository, and the models they provide either don't convert to onnx or are not readable by readNet().

What should be the next steps?

Reply:
> However, cv.dnn.readNet() wasn't able to read it 

The error was due to lack of support of dynamic input shape (dynamic H or W) I believe. You can use https://github.com/daquexian/onnx-simplifier to simplify it with a fixed input shape, such as [1, 3, 192, 192]. After that, OpenCV DNN should be able to load and infer the model.

Reply:
> However, cv.dnn.readNet() wasn't able to read it 

The error was due to lack of support of dynamic input shape (dynamic H or W) I believe. You can use https://github.com/daquexian/onnx-simplifier to simplify it with a fixed input shape, such as [1, 3, 192, 192]. After that, OpenCV DNN should be able to load and infer the model.

Reply:
I converted the v1 model using the command 
`onnxsim seg.onnx v1sim.onnx --overwrite-input-shape 1,3,192,192`

but now the onnx model is corrupted in some way : 
![Screenshot from 2023-02-17 15-32-43](https://user-images.githubusercontent.com/94868003/219615242-a7da1fef-bbad-4142-bf74-47413e8eb9fc.png)

Reply:
> I tried the Inference Model(softmax) of v1. It was converted to onnx successfully.

Did it really work? I downloaded the same model and use the same command as yours but the produced model has 0 Bytes.

Reply:
On [PaddleSeg](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.7/contrib/PP-HumanSeg#42-general-human-segmentation-models) for PP-HumanSegV1-Lite, the model downloaded with `checkpoint` link gives the error and produces a model of 0 Bytes.

But the model downloaded with `Inference Model (Softmax)` link is converted to onnx succesfully : 
![image](https://user-images.githubusercontent.com/94868003/219874704-3bea8e34-59eb-429d-ad46-3f452839b59b.png)
![image](https://user-images.githubusercontent.com/94868003/219874720-0e0dd8d2-5066-4f1c-93d8-75e54307538f.png)

I simplified the onnx  models produced by `Inference Model (Softmax)` and than tried the eval script on them.
These models were not read by `cv.dnn.readNet` which I shared above

Reply:
Thank you for pointing out the correct model and command. The conversion was done successfully and the simplifier worked as well. Latest OpenCV (4.7) can load and infer the model without any issues. See my screen shot below. You could have a try with my model: [humanseg.sim.onnx.zip](https://github.com/opencv/opencv_zoo/files/10779034/humanseg.sim.onnx.zip)


![image](https://user-images.githubusercontent.com/17219438/219999042-0b46ac98-7e0c-4263-9524-2915ea7a531a.png)

Reply:
Thanks, the error was fixed by removing the `self.outputNames` argument from `model.forward()`.

The eval script achieved the expected mIoU of 0.8602 on the model you shared above  (humanseg.sim.onnx) : 
![image](https://user-images.githubusercontent.com/94868003/220078200-d04dda58-3564-4f15-9bc6-267ca5b0dcda.png)

This matches the results from the PaddleSeg script : 
![image](https://user-images.githubusercontent.com/94868003/220078958-25458ce2-e9fa-4df9-92d6-3cd00d1addf8.png)
 
With the same script, for the model present in the zoo, the mIoU is 0.8996 : 
![image](https://user-images.githubusercontent.com/94868003/220079082-b248a1be-5256-4e00-a611-42928890d144.png)

Reply:
👍